File: lib/core/session/session_history.dart
import 'dart:convert';
import 'dart:typed_data';

import '../../src/rust/api/session_history.dart' as rust_history;
import '../../src/rust/api/sessions.dart' as rust_sess;
import 'session.dart';

/// Snapshot of session state for undo/redo.
///
/// Pure Dart value class — UI / Riverpod consumers read `sessions`,
/// `emptyFolders` and `description` directly. Persistence to the
/// Rust-side stack happens through `_encode` / `_decode` on the
/// [SessionHistory] boundary; the Rust actor stores opaque bytes.
class SessionSnapshot {
  final List<Session> sessions;
  final Set<String> emptyFolders;
  final String description;

  SessionSnapshot({
    required this.sessions,
    required this.emptyFolders,
    required this.description,
  });
}

/// Undo/redo history stack for session operations.
///
/// State machine ownership lives in `lfs_core::session_history`
/// (FRB sync). This Dart class is a thin handle wrapper:
/// `SessionHistory()` mints a Rust-side actor on construction;
/// method calls round-trip through FRB; the snapshot envelope
/// (`{sessions: [...], emptyFolders: [...], description: ...}`)
/// codec also lives Rust-side under
/// `lfs_core::session_json::{encode,decode}_snapshot_envelope` so
/// the Dart class never opens the blob.
///
/// Pair every `SessionHistory()` with a [dispose] when the owning
/// notifier tears down — otherwise the actor handle leaks for the
/// lifetime of the process.
class SessionHistory {
  late final BigInt _handleId;
  bool _disposed = false;

  SessionHistory() {
    _handleId = rust_history.sessionHistoryCreate();
  }

  bool get canUndo {
    if (_disposed) return false;
    return rust_history.sessionHistoryCanUndo(handleId: _handleId);
  }

  bool get canRedo {
    if (_disposed) return false;
    return rust_history.sessionHistoryCanRedo(handleId: _handleId);
  }

  String? get undoDescription {
    if (_disposed) return null;
    return rust_history.sessionHistoryUndoDescription(handleId: _handleId);
  }

  String? get redoDescription {
    if (_disposed) return null;
    return rust_history.sessionHistoryRedoDescription(handleId: _handleId);
  }

  /// Save current state before a destructive operation.
  void pushUndo(SessionSnapshot snapshot) {
    if (_disposed) return;
    rust_history.sessionHistoryPushUndo(
      handleId: _handleId,
      description: snapshot.description,
      blob: _encode(snapshot),
    );
  }

  /// Pop the last undo snapshot and push current state onto redo
  /// stack.
  SessionSnapshot? undo(SessionSnapshot currentState) {
    if (_disposed) return null;
    final result = rust_history.sessionHistoryUndo(
      handleId: _handleId,
      currentDescription: currentState.description,
      currentBlob: _encode(currentState),
    );
    if (result == null) return null;
    return _decode(result.blob, result.description);
  }

  /// Pop the last redo snapshot and push current state onto undo
  /// stack.
  SessionSnapshot? redo(SessionSnapshot currentState) {
    if (_disposed) return null;
    final result = rust_history.sessionHistoryRedo(
      handleId: _handleId,
      currentDescription: currentState.description,
      currentBlob: _encode(currentState),
    );
    if (result == null) return null;
    return _decode(result.blob, result.description);
  }

  void clear() {
    if (_disposed) return;
    rust_history.sessionHistoryClear(handleId: _handleId);
  }

  /// Drop the Rust-side actor. Idempotent. Call from the owning
  /// notifier's `dispose` so the per-handle state evicts from the
  /// process registry.
  void dispose() {
    if (_disposed) return;
    _disposed = true;
    rust_history.sessionHistoryDrop(handleId: _handleId);
  }

  static Uint8List _encode(SessionSnapshot snapshot) {
    // Plaintext credentials never enter the undo blob.
    // [SessionMutator] only ever stores `withoutCredentials()`
    // shapes in `state`, so the credential-bearing fields are
    // empty Strings; routing through `includeCredentials: false`
    // keeps the blob structurally credential-clean by construction.
    //
    // Both the per-session canonical encode and the outer
    // `{sessions, emptyFolders, description}` envelope wrap live
    // Rust-side in `lfs_core::session_json` — the Dart caller only
    // builds the typed input list and reads back the wire bytes.
    final inputs = snapshot.sessions
        .map((s) => sessionToJsonInput(s, includeCredentials: false))
        .toList(growable: false);
    final encoded = rust_sess.sessionHistoryEncodeSnapshotEnvelope(
      sessions: inputs,
      emptyFolders: snapshot.emptyFolders.toList(growable: false),
      description: snapshot.description,
    );
    return Uint8List.fromList(utf8.encode(encoded));
  }

  static SessionSnapshot _decode(Uint8List blob, String description) {
    final raw = utf8.decode(blob);
    final envelope = rust_sess.sessionHistoryDecodeSnapshotEnvelope(json: raw);
    final sessions = envelope.sessions.map(sessionFromJsonOutput).toList();
    return SessionSnapshot(
      sessions: sessions,
      emptyFolders: envelope.emptyFolders.toSet(),
      // The description in the envelope is informational only —
      // the registry hands the authoritative description back as a
      // sibling string argument so the Dart caller can preserve
      // its existing undo/redo menu-label contract.
      description: description,
    );
  }
}

File: rust/crates/lfs_frb/src/api/session_history.rs
//! FRB adapter for `lfs_core::session_history`. Per-handle
//! undo/redo stack — Dart serialises `SessionSnapshot` to bytes,
//! Rust stores opaque blobs + the description label.

#[derive(Debug, Clone)]
pub struct DbSessionHistorySnapshot {
    pub description: String,
    pub blob: Vec<u8>,
}

#[flutter_rust_bridge::frb(sync)]
pub fn session_history_create() -> u64 {
    lfs_core::session_history::create()
}

#[flutter_rust_bridge::frb(sync)]
pub fn session_history_drop(handle_id: u64) {
    lfs_core::session_history::drop_handle(handle_id);
}

#[flutter_rust_bridge::frb(sync)]
pub fn session_history_push_undo(handle_id: u64, description: String, blob: Vec<u8>) {
    lfs_core::session_history::push_undo(handle_id, description, blob);
}

#[flutter_rust_bridge::frb(sync)]
pub fn session_history_undo(
    handle_id: u64,
    current_description: String,
    current_blob: Vec<u8>,
) -> Option<DbSessionHistorySnapshot> {
    lfs_core::session_history::undo(handle_id, current_description, current_blob).map(|s| {
        DbSessionHistorySnapshot {
            description: s.description,
            blob: s.blob,
        }
    })
}

#[flutter_rust_bridge::frb(sync)]
pub fn session_history_redo(
    handle_id: u64,
    current_description: String,
    current_blob: Vec<u8>,
) -> Option<DbSessionHistorySnapshot> {
    lfs_core::session_history::redo(handle_id, current_description, current_blob).map(|s| {
        DbSessionHistorySnapshot {
            description: s.description,
            blob: s.blob,
        }
    })
}

#[flutter_rust_bridge::frb(sync)]
pub fn session_history_clear(handle_id: u64) {
    lfs_core::session_history::clear(handle_id);
}

#[flutter_rust_bridge::frb(sync)]
pub fn session_history_can_undo(handle_id: u64) -> bool {
    lfs_core::session_history::can_undo(handle_id)
}

#[flutter_rust_bridge::frb(sync)]
pub fn session_history_can_redo(handle_id: u64) -> bool {
    lfs_core::session_history::can_redo(handle_id)
}

#[flutter_rust_bridge::frb(sync)]
pub fn session_history_undo_description(handle_id: u64) -> Option<String> {
    lfs_core::session_history::undo_description(handle_id)
}

#[flutter_rust_bridge::frb(sync)]
pub fn session_history_redo_description(handle_id: u64) -> Option<String> {
    lfs_core::session_history::redo_description(handle_id)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn fresh_handle() -> u64 {
        let h = session_history_create();
        // Reset state in case a previous test left this handle dirty
        // (the registry is a process-wide singleton).
        session_history_clear(h);
        h
    }

    #[test]
    fn fresh_handle_can_neither_undo_nor_redo() {
        let h = fresh_handle();
        assert!(!session_history_can_undo(h));
        assert!(!session_history_can_redo(h));
        assert!(session_history_undo_description(h).is_none());
        assert!(session_history_redo_description(h).is_none());
        session_history_drop(h);
    }

    #[test]
    fn push_then_undo_round_trips_blob_and_description() {
        let h = fresh_handle();
        session_history_push_undo(h, "added session".into(), b"old-blob".to_vec());
        assert!(session_history_can_undo(h));
        assert_eq!(
            session_history_undo_description(h),
            Some("added session".to_string())
        );

        // Undo with a "current" snapshot — the registry pushes the
        // current state to the redo stack and pops the previous
        // entry off the undo stack.
        let undone = session_history_undo(h, "current".into(), b"new-blob".to_vec())
            .expect("undo yields previous snapshot");
        assert_eq!(undone.description, "added session");
        assert_eq!(undone.blob, b"old-blob");
        assert!(session_history_can_redo(h));
        assert!(!session_history_can_undo(h));
        session_history_drop(h);
    }

    #[test]
    fn redo_after_undo_round_trips_current_state() {
        let h = fresh_handle();
        session_history_push_undo(h, "step-1".into(), b"v1".to_vec());
        let _ = session_history_undo(h, "step-2".into(), b"v2".to_vec());
        // The stash from the undo (description="step-2", blob=b"v2")
        // is on the redo stack now. Redoing replays it.
        let redone = session_history_redo(h, "back-to-1".into(), b"v1".to_vec())
            .expect("redo yields stashed snapshot");
        assert_eq!(redone.description, "step-2");
        assert_eq!(redone.blob, b"v2");
        session_history_drop(h);
    }

    #[test]
    fn clear_drops_both_stacks() {
        let h = fresh_handle();
        session_history_push_undo(h, "x".into(), b"x".to_vec());
        session_history_clear(h);
        assert!(!session_history_can_undo(h));
        assert!(!session_history_can_redo(h));
        session_history_drop(h);
    }

    #[test]
    fn drop_handle_is_idempotent_on_unknown_id() {
        // Dart wrappers may call drop after a handle has already
        // been GC'd through the bus path; the call must be a no-op
        // rather than panic.
        session_history_drop(0xDEAD_BEEF);
    }
}
